File: python/heap-priority-queue/task-scheduler.py
import queue


class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        q = queue.PriorityQueue()
        lastseen = {}
        count = {}
        used = {}
        for task in tasks:
            lastseen[task] = -n
            count[task] = count.get(task, 0) + 1

        for task in count.keys():
            q.put((-count[task], task))

        finalcount = 0
        while len(q.queue) > 0:
            current = q.get()
            if current[0]:
                count = lastseen[current[1]] + n + 1
                while count in used:
                    count += 1
                lastseen[current[1]] = count
                used[count] = 1
                finalcount = max(finalcount, count)
                q.put((current[0] + 1, current[1]))

        return finalcount

File: c++/heap-priority-queue/task-scheduler.cc
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        std::priority_queue<std::pair<int, char>> pq;
        std::unordered_map<char, int> count;
        for (char c : tasks) {
            count[c] += 1;
            pq.push(std::make_pair(count[c], c));
        }

        int answer = 0;
        std::unordered_map<char, int> last;
        std::unordered_map<int, bool> taken;
        while (pq.size() != 0) {
            std::pair<int, char> current = pq.top();
            pq.pop();

            if (!last.contains(current.second)) {
                last[current.second] = -n;
            }

            int next = last[current.second] + n + 1;
            while (taken[next] && taken[next] == true) {
                next++;
            }
            taken[next] = true;
            last[current.second] = next;
            answer = std::max(answer, next);
        }

        return answer;
    }
};
